src/theory/quantifiers/sygus_grammar_norm.cpp

```diff
@@ -17,162 +17,472 @@
+                                                 const Datatype& dt)
 {
+  /* Use the sygus type to not lose reference to the original types (Bool,
+   * Int, etc) */
+  d_dt.setSygus(dt.getSygusType(),
+                sygus_norm->d_sygus_vars.toExpr(),
+                dt.getSygusAllowConst(),
+                dt.getSygusAllowAll());
+  for (unsigned i = 0, size_d_ops = d_ops.size(); i < size_d_ops; ++i)
+  {
+    d_dt.addSygusConstructor(d_ops[i].toExpr(),
+                             d_cons_names[i],
+                             d_cons_args_t[i],
+                             d_pc[i],
+                             d_weight[i]);
+  }
+  Trace("sygus-grammar-normalize") << "...built datatype " << d_dt << " ";
+  /* Add to global accumulators */
+  sygus_norm->d_dt_all.push_back(d_dt);
+  sygus_norm->d_unres_t_all.insert(d_unres_tn.toType());
+  Trace("sygus-grammar-normalize") << "---------------------------------\n";
 }
 
-/* recursion depth is limited by the height of the types, which is small  */
-void SygusGrammarNorm::collectInfoFor(TypeNode tn,
-                                      std::vector<TypeObject>& tos,
-                                      std::map<TypeNode, Type>& tn_to_unres,
-                                      std::map<TypeNode, bool>& visited)
+std::map<TypeNode, std::vector<Kind>> SygusGrammarNorm::TransfChain::d_chain_ops = {};
+std::map<TypeNode, std::map<Kind, Node>>
+    SygusGrammarNorm::TransfChain::d_chain_op_id = {};
+
+bool SygusGrammarNorm::TransfChain::isChainable(TypeNode tn, Node op)
 {
-  if (visited.find(tn) != visited.end())
+  std::vector<Kind> ops;
+  auto it = d_chain_ops.find(tn);
+  /* Creates map for type node */
+  if (it == d_chain_ops.end())
   {
-    return;
+    if (tn.isInteger())
+    {
+      ops = {kind::PLUS};
+    }
+    else
+    {
+      ops = {};
+    }
+    d_chain_ops[tn] = ops;
   }
-  visited[tn] = true;
-  Assert(tn.isDatatype());
-  /* Create new type name */
-  std::stringstream ss;
-  ss << tn << "_norm";
-  std::string type_name = ss.str();
-  /* Add to global accumulators */
-  tos.push_back(TypeObject(tn, type_name));
-  const Datatype& dt = static_cast<DatatypeType>(tos.back().d_t).getDatatype();
-  tn_to_unres[tn] = tos.back().d_unres_t;
-  /* Visit types of constructor arguments */
-  for (const DatatypeConstructor& cons : dt)
+  else
   {
-    for (const DatatypeConstructorArg& arg : cons)
+    ops = it->second;
+  }
+  /* TODO #1304: have more general ways of finding kind of operator, e.g. if op
+     is (\lambda xy. x + y) this function should realize that it is chainable
+     for integers */
+  /* Checks whether operator occurs in list of chainable ones */
+  return std::find(ops.begin(),
+                   ops.end(),
+                   NodeManager::currentNM()->operatorToKind(op))
+         != ops.end();
+}
+
+bool SygusGrammarNorm::TransfChain::isId(TypeNode tn, Kind op_k, Node elem)
```

Comment:
This function should also probably take Node op instead of Kind op_k so that we have a consistent interface wrt isChainable.